AppCenter/AppCenter/Internals/Context/AuthToken/MSAuthTokenContext.m

```diff
@@ -53,51 +48,75 @@ + (void)resetSharedInstance {
   sharedInstance = nil;
 }
 
-- (void)setAuthToken:(NSString *)authToken withAccountId:(NSString *)accountId {
+- (void)setAuthToken:(nullable NSString *)authToken withAccountId:(nullable NSString *)accountId expiresOn:(nullable NSDate *)expiresOn {
   NSArray *synchronizedDelegates;
-  BOOL isNewUser = NO;
+  BOOL isNewAccount = NO;
   @synchronized(self) {
-    self.authToken = authToken;
-    isNewUser = ![self.homeAccountId isEqualToString:accountId];
-    self.homeAccountId = accountId;
 
-    // Don't invoke the delegate while locking; it might be locking too and deadlock ourselves.
-    synchronizedDelegates = [self.delegates allObjects];
-  }
-  for (id<MSAuthTokenContextDelegate> delegate in synchronizedDelegates) {
-    if ([delegate respondsToSelector:@selector(authTokenContext:didSetAuthToken:)]) {
-      [delegate authTokenContext:self didSetAuthToken:authToken];
+    // If a nil authToken is passed with non-nil paarmeters, reset them.
+    if (!authToken) {
+      accountId = nil;
+      expiresOn = nil;
     }
-    if (isNewUser && [delegate respondsToSelector:@selector(authTokenContext:didUpdateAccountIdWithAuthToken:)]) {
-      [delegate authTokenContext:self didUpdateAccountIdWithAuthToken:authToken];
-    }
-  }
-}
+    NSMutableArray<MSAuthTokenInfo *> *authTokenHistory = [[self authTokenHistory] mutableCopy];
 
-- (BOOL)clearAuthToken {
-  NSArray *synchronizedDelegates;
-  BOOL clearedExistingUser = NO;
-  @synchronized(self) {
-    if (!self.authToken) {
-      return NO;
-    } else if (self.homeAccountId) {
-      clearedExistingUser = YES;
+    MSAuthTokenInfo *lastEntry = authTokenHistory.lastObject;
+    NSString *__nullable latestAuthToken = lastEntry.authToken;
+    NSString *__nullable latestAccountId = lastEntry.accountId;
+    NSDate *__nullable latestTokenEndTime = lastEntry.expiresOn;
+
+    // If new token doesn't differ from the last token of array - no need to add it to array.
+    if (lastEntry != nil && [latestAuthToken isEqual:(NSString * _Nonnull) authToken]) {
+      return;
+    }
+    BOOL isNewUser = authTokenHistory.lastObject == nil || ![accountId isEqualToString:(NSString * __nonnull) latestAccountId];
+    NSDate *newTokenStartDate = [NSDate date];
+
+    // If there is a gap between tokens.
+    if (latestTokenEndTime && [newTokenStartDate laterDate:(NSDate * __nonnull) latestTokenEndTime]) {
+
+      // If the account is the same or becomes anonymous.
+      if (!isNewUser || authToken == nil) {
+
+        // Apply the new token to this time.
+        newTokenStartDate = latestTokenEndTime;
```

Comment:
Should it be
```suggestion
        newTokenStartDate = [NSDate new];
```

Start time for the new token is not the end time of previous token.

Reply:
Discussed offline. Seems correct.